API/SSW.Rewards.Application/Users/Commands/UploadProfilePic/UploadProfilePicCommand.cs

```diff
@@ -3,51 +3,81 @@
+        public ICurrentUserService _currentUserService { get; }
 
-        public class UploadProfilePicHandler : IRequestHandler<UploadProfilePicCommand, string>
+        public UploadProfilePicHandler(ICurrentUserService currentUserService,
+            ISSWRewardsDbContext sSWRewardsDbContext,
+            IProfilePicStorageProvider profilePicStorageProvider,
+            IMapper mapper)
         {
-            private readonly IMapper _mapper;
-            private readonly ISSWRewardsDbContext _context;
-            private readonly IProfilePicStorageProvider _storage;
-            public ICurrentUserService _currentUserService { get; }
-
-            public UploadProfilePicHandler(ICurrentUserService currentUserService,
-                ISSWRewardsDbContext sSWRewardsDbContext,
-                IProfilePicStorageProvider profilePicStorageProvider,
-                IMapper mapper)
-            {
-                _currentUserService = currentUserService;
-                _mapper = mapper;
-                _storage = profilePicStorageProvider;
-                _context = sSWRewardsDbContext;
-            }
+            _currentUserService = currentUserService;
+            _mapper = mapper;
+            _storage = profilePicStorageProvider;
+            _context = sSWRewardsDbContext;
+        }
 
-            public async Task<string> Handle(UploadProfilePicCommand request, CancellationToken cancellationToken)
-            {
-	            await using var ms = new MemoryStream();
-				IFormFile file = request.File;
-				await file.CopyToAsync(ms, cancellationToken);
+        public async Task<ProfilePicResponseDto> Handle(UploadProfilePicCommand request, CancellationToken cancellationToken)
+        {
+            var response = new ProfilePicResponseDto();
 
-				byte[] bytes = ms.ToArray();
+            await using var ms = new MemoryStream();
 
-				string filename = file.FileName;
+            IFormFile file = request.File;
+            
+            await file.CopyToAsync(ms, cancellationToken);
+
+            byte[] bytes = ms.ToArray();
+
+            string filename = file.FileName;
+
+            string url = await _storage.UploadProfilePic(bytes, filename);
+            
+            var user = await _context.Users
+                .Where(u => u.Email.ToLower() == _currentUserService.GetUserEmail())
+                .Include(u => u.UserAchievements)
+                    .ThenInclude(ua => ua.Achievement)
+                .AsNoTracking()
+                .FirstOrDefaultAsync(cancellationToken);
+            
+            user.Avatar = url;
+
+            var achievement = await _context.Achievements.FirstOrDefaultAsync(a => a.Name == MilestoneAchievements.ProfilePic, cancellationToken);
+
+            if (!user.UserAchievements.Any(a => a.Achievement == achievement))
+            {
+                user.UserAchievements.Add(new UserAchievement
+                {
+                    Achievement = achievement
+                });
+
```

Comment:
When you get the user on line 55, you're doing it AsNoTracking(). Then you're modifying the user by adding this achievement. Isn't that wrong?

Reply:
Done